### PR TITLE
Messages download

### PR DESCRIPTION
**Pull request by almirsarajcic**
@sergey-lukianov, can you please check if I'm heading in the right direction?

~~When the user selects messages we go through the files, create an archive in the memory, and put the files inside.
Then we redirect the user to the file URL.~~

https://user-images.githubusercontent.com/845364/205993057-73b3ec29-3001-4351-b20f-36b8f0e81109.mov

~~While working on this I've found two ways of letting the user download it:~~
1. ~~create the chunk, save the file in the DB, and redirect the user to the file URL~~
2. ~~save the archive in the static directory and redirect the user to the static path~~

~~If we proceed with saving an archive in the DB, we'll need to create multiple chunks as currently there's an issue if the archive is larger than 10 MB (see the comments).~~

TODO:
- [x] Enable download in Chat (Dialog)
- [x] Enable download in Room

### CODE REVIEW

**Comment by sergey-lukianov** [comment]
@almirsarajcic  The feature ideal outcome is zip archive with 
- `index.html` - decrypted selected messages rendered as in chat (linked to `files/`)
- `files/` dir to hold decrypted file attachments

Tech design
- frontend passes ids of messages to render
- LiveView saves decryption key, message ids and dialog key in `Broker`
- User is redirected to ZipController having broker_key as url param
- ZipController reads out dialog messages and files. Generates streamed zip archive (with something like Zstream)

We never store decrypted content

Zstream - https://hexdocs.pm/zstream/readme.html

**Comment by almirsarajcic** [comment]
Thanks, @sergey-lukianov! 👍

Here's the new approach:
https://user-images.githubusercontent.com/845364/206547469-37c6d462-88d3-4b56-8c15-13119ee37cf2.mov

It currently only works for the dialog, but tomorrow I'll enable it in the rooms as well, and add the finishing touches.

File: lib/chat_web/controllers/zip_controller.ex [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+
+      file_entries =
+        messages
+        |> Enum.reduce([], fn message, file_entries ->
+          with %{type: type, content: json} when type in [:file, :image, :video] <- message,
+               {id, content} <- StorageId.from_json(json),
+               [chunk_key, chunk_secret_raw, _, _type, filename, _size] <- Files.get(id, content),
+               chunk_secret <- Base.decode64!(chunk_secret_raw),
+               file_stream <- ChunkedFiles.stream_chunks(chunk_key, chunk_secret) do
+            entry = Zstream.entry("files/#{filename}", file_stream)
```
Comment (sergey-lukianov) [inline lib/chat_web/controllers/zip_controller.ex:79]: `filename` comes from user and may conflict when put a bunch of them in a dir.
Maybe add suffix to filename

File: lib/chat_web/controllers/zip_controller.ex [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+      index_entry = Zstream.entry("index.html", exported_messages)
+
+      file_entries =
+        messages
+        |> Enum.reduce([], fn message, file_entries ->
+          with %{type: type, content: json} when type in [:file, :image, :video] <- message,
+               {id, content} <- StorageId.from_json(json),
+               [chunk_key, chunk_secret_raw, _, _type, filename, _size] <- Files.get(id, content),
+               chunk_secret <- Base.decode64!(chunk_secret_raw),
+               file_stream <- ChunkedFiles.stream_chunks(chunk_key, chunk_secret) do
```
Comment (sergey-lukianov) [inline lib/chat_web/controllers/zip_controller.ex:78]: `file_stream` is a binary with whole file, not a stream. This needs to be stream to make memory footprint not to reflect file size. Files could be huge. Target system may have 2G of physical RAM only

File: lib/chat_web/controllers/zip_controller.ex [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+         {messages_ids, user_data, user_id, timezone} <- Broker.get(broker_key),
+         {identity, _rooms} = User.device_decode(user_data),
+         peer <- User.by_id(user_id),
+         dialog <- Dialogs.find_or_open(identity, peer),
+         {messages, users} when not is_nil(messages) and not is_nil(users) <-
+           fetch_messages_and_users_ids(dialog, identity, messages_ids) do
+      conn =
+        conn
+        |> put_resp_content_type("application/zip")
+        |> put_resp_header("content-disposition", "attachment; filename=messages.zip")
```
Comment (sergey-lukianov) [inline lib/chat_web/controllers/zip_controller.ex:33]: Neat: may be introduce some naming pattern so user will know what is in it

File: lib/chat_web/controllers/zip_controller.ex (hunk 2 above)
Comment (sergey-lukianov) [reply lib/chat_web/controllers/zip_controller.ex:78]: File chunk go back to db. So there would be `CubDB.Select` to get a scream of chunks.
Check the https://github.com/Buckitup-chat/chat/pull/63 or wait till it gets into master. Btw, any critics is appreciative

**Comment by sergey-lukianov** [comment]
Overall looks great.
I do not see a clear reason to make `ExportedMessage` out of `Message`. Lets finish current functionality and then discuss

**Comment by almirsarajcic** [comment]
The reason why I've extracted the `ExportedMessage` component is that these messages have a header without the options, but also the files are shown differently.

I've reduced the amount of code in it, though, as I'm now relying on the original `Message` component for rendering other types of messages that don't contain files.

File: lib/chat_web/controllers/zip_controller.ex [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+               {id, content} <- StorageId.from_json(json),
+               [chunk_key, chunk_secret_raw, _, _type, filename, _size] <- Files.get(id, content),
+               chunk_secret <- Base.decode64!(chunk_secret_raw),
+               file_stream <- ChunkedFiles.stream_chunks(chunk_key, chunk_secret) do
+            {extension, filename} =
+              filename
+              |> String.split(".")
+              |> List.pop_at(-1)
+
+            filename = Enum.join(filename, ".") <> "_" <> id <> "." <> extension
```
Comment (sergey-lukianov) [inline lib/chat_web/controllers/zip_controller.ex:80]: Reuse this in message render as well

File: lib/chat_web/live/main_live/layout/exported_message.ex [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+  defp assign_file(%{message: %{content: json}} = assigns) do
+    {id, secret} = StorageId.from_json(json)
+    [_, _, _, _, name, size] = Files.get(id, secret)
+
+    {extension, filename} =
+      name
+      |> String.split(".")
+      |> List.pop_at(-1)
+
+    filename = Enum.join(filename, ".") <> "_" <> id <> "." <> extension
```
Comment (sergey-lukianov) [inline lib/chat_web/live/main_live/layout/exported_message.ex:124]: copy-paste detected

File: lib/chat_web/live/main_live/page/room.ex [hunk 6]
```diff
@@ -279,6 +280,22 @@ defmodule ChatWeb.MainLive.Page.Room do
...
+    messages_ids =
+      messages
+      |> Jason.decode!()
+      |> Enum.map(fn %{"id" => message_id, "index" => index} ->
+        {String.to_integer(index), message_id}
+      end)
+
+    key = Broker.store({:room, {messages_ids, room, room_identity}, timezone})
+
+    push_event(socket, "chat:redirect", %{url: Routes.zip_url(socket, :get, key)})
```
Comment (sergey-lukianov) [inline lib/chat_web/live/main_live/page/room.ex:296]: verified routes here too, please

**Review by sergey-lukianov — CHANGES_REQUESTED** [review]

File: lib/chat_web/live/main_live/layout/exported_message.ex [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+
+  alias Chat.Files
+  alias Chat.Identity
+  alias Chat.Memo
+  alias Chat.RoomInvites
+  alias Chat.Utils
+  alias Chat.Utils.StorageId
+  alias ChatWeb.MainLive.Layout.Message
+
+  def message_block(assigns) do
```
[the remark this answers is not in the document]
Comment (sergey-lukianov) [reply lib/chat_web/live/main_live/layout/exported_message.ex:12]: I see that text messages are reused
May we do so for files as well ?

File: lib/chat_web/live/main_live/layout/message.ex [hunk 8]
```diff
@@ -1,68 +1,151 @@
...
-      _ -> render_text(assigns)
-    end
+  def message_block(assigns) do
+    assigns =
+      assigns
+      |> assign_new(:export?, fn -> false end)
+      |> assign_new(:room, fn -> nil end)
+      |> assign_new(:is_mine?, fn
+        %{export?: true} ->
+          false
```
[the remark this answers is not in the document]
Comment (almirsarajcic) [reply lib/chat_web/live/main_live/layout/message.ex:29]: I assumed the downloaded messages should look the same regardless of who downloaded them. I'll fix that.

**Review by sergey-lukianov — APPROVED** [review]

**Pull request merged**